---
title: "Rapport phylo-géographique"
date: "`r Sys.Date()`"
output: pdf_document
params:
  OutputDirectory: "/home/virologie/pipeline_stage/"
  Reference: "/temp.txt"
---

```{=tex}
\section{Introduction}

Ce rapport contient des informations vis à vis de la qualité et des résultats d'un séquençage associé à sa séquence de référence.
```
```{r, include = FALSE}
#General settings

knitr::opts_knit$set(root.dir = getwd())

knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)

```

```{r}
#Download libraries
library(knitr)
library(seqinr)
library(dplyr)
library(scales)
library(ggplot2)
library(ggpubr)
library(tidyr)
library(stringr)
library(microseq)
library(spatstat)
library(kableExtra)
source("./functionPlots.R")
source("./functionPlotsMulti.R")
```

```{r}
# Load data #
OutputDirectory <- params$OutputDirectory
reference <- params$Reference

listOutputs <- list.dirs(OutputDirectory, full.names = TRUE, recursive = FALSE)
fastqDataList <- list()
samDataList <- list()
coverageList <- list()
seq_consensusList <- list()
variantList <- list()

fastqData <- list(data.frame(GC_content = integer(), occurence = integer(), patient = character()),
                  data.frame(read_length = integer(), occurence = integer(), patient = character()),
                  data.frame(Seq_score = integer(), occurence = integer(), patient = character())) 

samData <- list(data.frame(read = character(), occurence = integer(), patient = character()),
                data.frame(Map_score = integer(), occurence = integer(), patient = character()),
                data.frame(read_length = integer(), mapped = integer(), occurence = integer(), patient = character()),
                data.frame(Seq_score = integer(), occurence = integer(), patient = character()))

variant <- data.frame(V1 = character(), V2 = integer(), V3 = character(),
                      V4 = character(), V5 = character(), V6 = integer(), 
                      V7 = character(), V8 = character(), V9 = character(), 
                      V10 = character(), patient = character())

patientNumber <- length(listOutputs)
patientList <- list()
for (outputFolder in listOutputs) {
  
  patientList[[length(patientList)+1]] <- basename(outputFolder)
  
  ## Fastq sequence file ##
  fastqDataList[[length(fastqDataList)+1]]<- 
    addPatientID(getAllFastqParseTables(outputFolder), basename(outputFolder))
  
  ## Sam pattern files ##
  samDataList[[length(samDataList)+1]]  <- 
    addPatientID(getAllSamParseTables(outputFolder), basename(outputFolder))
  
  ## Coverage file ##
  coverageList[[length(coverageList)+1]]  <- 
    read.table(paste0(outputFolder, "/results/Coverage.bed"), sep = "\t")%>%  mutate(patient = basename(outputFolder))
  
  colnames(coverageList[[length(coverageList)]])  <- c("sample", "start", "stop", "depth", "patient")
  
 
  ## consensus sequence ##
  seq_consensusList[[length(seq_consensusList)+1]]  <- 
    list(read.fasta(paste0(outputFolder, "/results/Consensus_Sequence/consensus.fa"), as.string = TRUE , seqonly = T), basename(outputFolder))


  ## variant VCF file ###
  variantList[[length(variantList)+1]]  <- 
    read.table(paste0(outputFolder, "/python_vcf_result.vcf"), sep = "\t") %>% mutate(patient = basename(outputFolder))
}

fastqData <- combineFastqTables(fastqDataList)
samData <- combineSamTables(samDataList)
variant <- combineVariantTables(variantList)
```

```{=tex}
\section{1. Qualité du séquençage}

```

```{r}
fastqGCTab <- data.frame(patient = character(), Moyenne = double(),
                            Mediane = double(), ecarttype = double())

fastqLengthTab <- data.frame(patient = character(), Moyenne = double(),
                            Mediane = double())

fastqScoreTab <- data.frame(patient = character(), Moyenne = double(),
                            Mediane = double())

for(p in patientList){
  fastqGCTab <- fastqGCTab %>% bind_rows(fastqData[[1]] 
                      %>% filter(patient == p) %>% group_by(patient) 
                      %>% summarise(Moyenne = weighted.mean( GC_content, occurence),
                          Mediane = weighted.median(GC_content, occurence), 
                          ecarttype = sd(GC_content)))
  fastqLengthTab <- fastqLengthTab %>% bind_rows(fastqData[[2]] 
                      %>% filter(patient == p) %>% group_by(patient) 
                      %>% summarise(Moyenne = weighted.mean(read_length, occurence),Mediane = weighted.median(read_length, occurence)))
  
  fastqScoreTab <- fastqScoreTab %>% bind_rows(fastqData[[3]] 
                      %>% filter(patient == p) %>% group_by(patient) 
                      %>% summarise(Moyenne = weighted.mean(Seq_score, occurence),Mediane = weighted.median(Seq_score, occurence)))
}
if(patientNumber <= 10){
  plots <- fastqQualityPlot(fastqGCTab, fastqLengthTab, fastqScoreTab)
  ggarrange(plots[[1]], plots[[2]], plots[[3]])
}
fastqGCTab %>% kable(caption = "Informations sur le score qualité des patients") %>% kable_paper(latex_options = c("hold_position"))
```

\pagebreak
```{=tex}
\section{2. Qualité du mapping sur la référence}

Cette section traite de données de qualités générales du mapping des lectures

```
```{r fig.height = 7, fig.width=7}

samMapTable <- data.frame(patient = character(),percent = double(), read = character())


samMapScoreTab <- data.frame(patient = character(), Moyenne = double(),
                            Mediane = double())

samScoreTab <- data.frame(patient = character(), Moyenne = double(),
                            Mediane = double())


for(p in patientList){
  
  samMapTable <- samMapTable %>% bind_rows(samData[[1]] %>% filter(patient == p) 
                 %>% relocate(patient, .before = read)  
                 %>% filter(read != "total") 
                 %>% group_by(patient) 
                 %>% reframe(percent = round(occurence / sum(occurence) * 100, digits = 2), read) 
  )
  
  samMapScoreTab <- samMapScoreTab %>% bind_rows(samData[[2]]
                      %>% filter(patient == p) %>% group_by(patient) 
                      %>% summarise(Moyenne = weighted.mean(Map_score, occurence),
                          Mediane = weighted.median(Map_score, occurence)))
  
  samScoreTab <- samScoreTab %>% bind_rows(samData[[4]]
                      %>% filter(patient == p) %>% group_by(patient) 
                      %>% summarise(Moyenne = weighted.mean(Seq_score, occurence),
                          Mediane = weighted.median(Seq_score, occurence)))
}

if(patientNumber <= 10){
  plots <- samMultiPlots(samMapTable, samMapScoreTab, samScoreTab)
  ggarrange(plots[[1]], plots[[2]], plots[[3]])
}

samMapTable[,c("patient","percent")] %>% kable(caption = "Informations sur le nombre de reads mappés par patients") %>% kable_paper(latex_options = c("hold_position"))
samMapScoreTab %>% kable(caption = "Informations sur le score d'alignement par patient") %>% kable_paper(latex_options = c("hold_position"))
samScoreTab %>% kable(caption = "Informations sur le score qualité des séquences mappées par patient") %>% kable_paper(latex_options = c("hold_position"))

```

```{r}
plotReadMapLength(samData[[3]])
```

\pagebreak
```{=tex}
\subsection{2.1. Couverture et profondeur du séquençage}

Cette section traite de la profondeur et de la couverture du séquençage

```
```{r}
##Coverage plot
plotCoverage(coverage)
```

```{=tex}
\subsection{2.2. Régions séquencées indéterminées}

Ce graphique montre en rouge les régions indeterminées, c'est à dire couverte par aucune séquence

```
```{r fig.height = 2, fig.width = 5}
# Rectangular plot that shows regions with 0 coverage
unknownRegions(seq_consensus)
```

\pagebreak
```{=tex}
\section{3. Analyse du taux de GC}

Cette section analyse le taux de GC des lectures et les compare à celui de la référence

```
```{r}
plots <- plotGC(fastqData[[1]], reference)  
plots[[1]]
```

```{r}
plots[[2]] %>% kable(caption = "Statistics comparison between reads and reference") %>% kable_paper(latex_options = c("hold_position"))
```

```{r}
plots[[3]] %>% kable(caption = "Kolmogorov-Smirnov test results") %>% kable_paper(latex_options = c("hold_position"))
```

```{r}
plots[[4]] %>% kable(caption = "Stats from Student test") %>% kable_paper(latex_options = c("hold_position"))
```

```{r}
plots[[5]] %>% kable(caption = "Confidence interval from Student test") %>% kable_paper(latex_options = c("HOLD_position"))
```

\pagebreak
```{=tex}
\section{4. Variants détectés}

Cette section est composée d'un graphe affichant les lieux ou des variants ont 
été trouvés, acompagné d'un tableau contenant de nombreux détails


```
```{r fig.height = 2, fig.width = 8}
content <- vcfAnalysis(variant, reference) 
content[[2]] 
```

```{r}
content[[1]] %>% kable(caption = "Variants detected") %>% kable_paper(latex_options = c("hold_position"))
```

```{=tex}
\section{5. Résultats sur plusieurs patients}

Cette section contient des informations sur les résultats sur plusieurs patients

```
```{=tex}
\subsection{Tableaux de données}

Cette section est composée d'un graphe affichant les lieux ou des variants ont 
été trouvés, acompagné d'un tableau contenant de nombreux détails


```
